File: netlify/functions/run_manager.py
"""POST /api/manager — General Manager routes task to relevant agents."""

import json
import os
import time
from datetime import datetime, timezone
from pathlib import Path

ROOT = Path(__file__).parent.parent.parent
CONFIG_PATH = ROOT / "config" / "agents.json"

CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type",
    "Access-Control-Allow-Methods": "POST, OPTIONS",
    "Content-Type": "application/json",
}

MODEL = "claude-sonnet-4-6"

ROUTING_KEYWORDS = {
    "linkedin": ["linkedin", "b2b", "professional", "réseau pro", "connexion"],
    "facebook": ["facebook", "meta", "fb", "communauté", "community"],
    "instagram": ["instagram", "reels", "story", "stories", "visuel", "visual"],
    "tiktok": ["tiktok", "viral", "short video", "courte vidéo", "trending"],
    "seo": ["seo", "référencement", "search", "google", "keyword", "mot-clé"],
    "content": ["blog", "article", "contenu", "content", "copywriting", "rédac"],
    "analytics": ["analytics", "statistiques", "kpi", "data", "rapport", "report"],
    "email": ["email", "mail", "newsletter", "campagne", "automation", "drip"],
}


def _load_agents() -> list[dict]:
    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        return json.load(f).get("agents", [])


def _route_task(task: str, agents: list[dict]) -> list[dict]:
    task_lower = task.lower()
    matched = []
    for agent in agents:
        role_lower = agent["role"].lower()
        for role_key, keywords in ROUTING_KEYWORDS.items():
            if role_key in role_lower:
                if any(kw in task_lower for kw in keywords):
                    matched.append(agent)
                    break
    # If no keyword match, pick top 3 most relevant by role order
    if not matched:
        matched = agents[:3]
    return matched[:4]  # max 4 agents per manager task


def _call_claude(system: str, user_msg: str) -> str:
    api_key = os.environ.get("ANTHROPIC_API_KEY", "").strip()
    if not api_key:
        return (
            "⚠️ **No API key configured.** Set ANTHROPIC_API_KEY in Netlify "
            "environment variables to enable the Silver Team to work."
        )
    try:
        import anthropic
        client = anthropic.Anthropic(api_key=api_key)
        response = client.messages.create(
            model=MODEL,
            max_tokens=3000,
            system=system,
            messages=[{"role": "user", "content": user_msg}],
        )
        return response.content[0].text
    except Exception as exc:
        return f"❌ API error: {exc}"


def handler(event, context):
    if event.get("httpMethod") == "OPTIONS":
        return {"statusCode": 204, "headers": CORS_HEADERS, "body": ""}

    if event.get("httpMethod") != "POST":
        return {"statusCode": 405, "headers": CORS_HEADERS, "body": json.dumps({"error": "Method not allowed"})}

    try:
        body = json.loads(event.get("body") or "{}")
        task = body.get("task", "").strip()
        if not task:
            return {
                "statusCode": 400,
                "headers": CORS_HEADERS,
                "body": json.dumps({"error": "task is required"}),
            }

        agents = _load_agents()
        routed_agents = _route_task(task, agents)

        # Build manager system prompt
        team_summary = "\n".join(
            f"- {a['name']} ({a['role']}): {', '.join(a['skills'][:4])}"
            for a in agents
        )
        manager_system = (
            "You are the General Manager of the Silver Team, an elite digital marketing agency. "
            "Your team members are:\n\n"
            f"{team_summary}\n\n"
            "When given a task, you:\n"
            "1. Analyse the request and identify which team members are most relevant\n"
            "2. Provide a strategic breakdown of how to approach it\n"
            "3. Delegate specific sub-tasks to team members with clear instructions\n"
            "4. Give a consolidated strategic recommendation\n\n"
            "Be decisive, strategic, and actionable. Respond in the same language as the task."
        )

        routed_names = [a["name"] for a in routed_agents]
        delegation_prompt = (
            f"Task to manage: {task}\n\n"
            f"Relevant team members identified: {', '.join(routed_names)}\n\n"
            "Provide:\n"
            "1. **Strategic overview** (2-3 sentences)\n"
            "2. **Delegation plan** — specific instructions for each relevant team member\n"
            "3. **Success metrics** — how to measure the outcome\n"
            "4. **Timeline** — realistic execution timeline"
        )

        start = time.monotonic()
        result = _call_claude(manager_system, delegation_prompt)
        elapsed_ms = int((time.monotonic() - start) * 1000)

        return {
            "statusCode": 200,
            "headers": CORS_HEADERS,
            "body": json.dumps({
                "task": task,
                "routed_to": routed_names,
                "result": result,
                "response_time_ms": elapsed_ms,
                "timestamp": datetime.now(timezone.utc).isoformat(),
            }),
        }
    except Exception as exc:
        return {
            "statusCode": 500,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": str(exc)}),
        }

File: netlify/functions/run_manager.js
// POST /api/manager — General Manager routes task to relevant agents
// Frontend sends { task, agents } — no file system access needed.

const CORS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Headers": "Content-Type",
  "Access-Control-Allow-Methods": "POST, OPTIONS",
  "Content-Type": "application/json",
};

const ROUTING_KEYWORDS = {
  linkedin:  ["linkedin", "b2b", "professionnel", "professional", "réseau pro"],
  facebook:  ["facebook", "meta", "fb", "communauté", "community"],
  instagram: ["instagram", "reels", "story", "stories", "visuel", "visual"],
  tiktok:    ["tiktok", "viral", "short video", "courte vidéo", "trending"],
  seo:       ["seo", "référencement", "search", "google", "keyword", "mot-clé"],
  content:   ["blog", "article", "contenu", "content", "copywriting", "rédac"],
  analytics: ["analytics", "statistiques", "kpi", "data", "rapport", "report"],
  email:     ["email", "mail", "newsletter", "campagne", "automation"],
};

function routeTask(task, agents) {
  const lower = task.toLowerCase();
  const matched = agents.filter((a) => {
    const roleKey = a.role.toLowerCase();
    return Object.entries(ROUTING_KEYWORDS).some(
      ([key, words]) => roleKey.includes(key) && words.some((w) => lower.includes(w))
    );
  });
  return matched.length ? matched.slice(0, 4) : agents.slice(0, 3);
}

async function callClaude(system, userMsg, apiKey) {
  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01",
    },
    body: JSON.stringify({
      model: "claude-sonnet-4-6",
      max_tokens: 3000,
      system,
      messages: [{ role: "user", content: userMsg }],
    }),
  });

  if (!res.ok) {
    const err = await res.text();
    throw new Error(`Anthropic API ${res.status}: ${err.slice(0, 200)}`);
  }
  const data = await res.json();
  return data.content[0].text;
}

exports.handler = async (event) => {
  if (event.httpMethod === "OPTIONS") {
    return { statusCode: 204, headers: CORS, body: "" };
  }
  if (event.httpMethod !== "POST") {
    return { statusCode: 405, headers: CORS, body: JSON.stringify({ error: "Method not allowed" }) };
  }

  try {
    const body = JSON.parse(event.body || "{}");
    const { task, agents = [] } = body;

    if (!task) {
      return { statusCode: 400, headers: CORS, body: JSON.stringify({ error: "task is required" }) };
    }

    const apiKey = process.env.ANTHROPIC_API_KEY || "";
    if (!apiKey) {
      return {
        statusCode: 200,
        headers: CORS,
        body: JSON.stringify({
          task,
          routed_to: [],
          result: `⚠️ **Clé API manquante.** Ajoute \`ANTHROPIC_API_KEY\` dans Netlify → Site settings → Environment variables, puis redéploie.`,
          success: false,
          response_time_ms: 0,
          timestamp: new Date().toISOString(),
        }),
      };
    }

    const routed = routeTask(task, agents);
    const routedNames = routed.map((a) => a.name);

    const teamSummary = agents
      .map((a) => `- ${a.name} (${a.role}): ${(a.skills || []).slice(0, 4).join(", ")}`)
      .join("\n");

    const system = [
      "You are the General Manager of the Silver Team, an elite digital marketing agency.",
      "Your team members are:\n" + teamSummary,
      "",
      "When given a task, you:",
      "1. Analyse the request and identify which team members are most relevant",
      "2. Provide a strategic breakdown of how to approach it",
      "3. Delegate specific sub-tasks to team members with clear instructions",
      "4. Give a consolidated strategic recommendation",
      "",
      "Be decisive, strategic, and actionable. Respond in the same language as the task.",
    ].join("\n");

    const userMsg = [
      `Task to manage: ${task}`,
      `Relevant team members identified: ${routedNames.join(", ")}`,
      "",
      "Provide:",
      "1. **Strategic overview** (2-3 sentences)",
      "2. **Delegation plan** — specific instructions for each relevant team member",
      "3. **Success metrics** — how to measure the outcome",
      "4. **Timeline** — realistic execution timeline",
    ].join("\n");

    const start = Date.now();
    const result = await callClaude(system, userMsg, apiKey);
    const elapsed = Date.now() - start;

    return {
      statusCode: 200,
      headers: CORS,
      body: JSON.stringify({
        task,
        routed_to: routedNames,
        result,
        success: true,
        response_time_ms: elapsed,
        timestamp: new Date().toISOString(),
      }),
    };
  } catch (err) {
    return {
      statusCode: 500,
      headers: CORS,
      body: JSON.stringify({ error: err.message }),
    };
  }
};
